### DISCUSSION TITLE
[lodash] `isObjectLike` and `isPlainObject` missing user-defined type guard

### DISCUSSION THREAD

**Discussion by otomad** [DefinitelyTyped/DefinitelyTyped#68867] [Issues with a @types package]
#33228 Add the type guard for `isObject`.

However, there is no type guard for `isObjectLike` and `isPlainObject`.

I want to use it for is an object, not null and function. And found that there is no type guard for `isObjectLike`.

just add them like
```typescript
isObjectLike(value?: any): value is object;

isPlainObject(value?: any): value is object;
```

[a comment by an automated account is not shown]

**Comment by aj-r** [comment]
The problem is that `isPlainObject` returns false for some objects that do have type `object`. So a type guard isn't feasible since the inverse type guard would be inaccurate.

`isObjectLike` is a little more compelling... I think this one should have the type guard, since its documentation aligns with what typescript considers an "object". 

In fact, `isObject` should probably NOT have a type guard, since it returns true for things like functions. But I think removing the type guard now would mainly just create unnecessary churn in people's code, so probably not worth it. Maybe we could change it to `value is object | function(...any[]): any` which is a little more accurate.

Feel free to send a PR.

**Reply by otomad** [reply]
`isPlainObject`: in `if()` can be constrain to an object. At least this function can ensure that it is an object when returns true.

`isObject`: maybe `value is object | Function` is also considerable?

**Comment by aj-r** [comment]
`isPlainObject`: unfortunately it's not possible to have a typeguard affect `if(isPlainObject())` without also affecting the `else` or `if(!isPlainObject())`. If we did this, the negated versions would just be wrong. I think users who want a type guard should use `isObject`/`isObjectLike` instead.

`isObject`: yes that could work, too. My understanding is that the `Function` type should be avoided, although I think your version is technically more correct.
